File: wisp/wisp-cli/Sources/wisp/Output.swift
import Foundation
import WispAI

// MARK: - Text mode (streaming)

/// Streams text deltas to stdout as they arrive.
/// Prints a trailing newline after the final token and exits with 0.
/// On error exits with 1.
func runTextMode(
    stream: AsyncStream<AssistantMessageEvent>
) async -> Int32 {
    var hadText = false
    for await event in stream {
        switch event {
        case .textDelta(_, let delta, _):
            print(delta, terminator: "")
            flushStdout()
            hadText = true
        case .thinkingDelta(_, let delta, _):
            // Print thinking to stderr so it doesn't pollute stdout pipeline.
            fputs(delta, stderr)
        case .done:
            if hadText { print() } // trailing newline
            return 0
        case .error(let msg):
            let desc = msg.errorMessage ?? "unknown error"
            fputs("\nError: \(desc)\n", stderr)
            if let diag = msg.diagnostics.first {
                fputs("  [\(diag.type)] \(diag.error?.message ?? "")\n", stderr)
            }
            return 1
        default:
            break
        }
    }
    return 0
}

// MARK: - JSON mode (newline-delimited)

/// Emits one JSON object per line to stdout for every event.
/// Mirrors the --mode json output of the original pi coding-agent.
func runJsonMode(
    stream: AsyncStream<AssistantMessageEvent>
) async -> Int32 {
    var exitCode: Int32 = 0
    for await event in stream {
        if let line = encodeEvent(event) {
            print(line)
            flushStdout()
        }
        switch event {
        case .error: exitCode = 1
        default: break
        }
    }
    return exitCode
}

// MARK: - Event → JSON

private func encodeEvent(_ event: AssistantMessageEvent) -> String? {
    var dict: [String: Any]
    switch event {
    case .start(let partial):
        dict = ["type": "start", "model": partial.model, "provider": partial.provider]

    case .textStart(let idx, _):
        dict = ["type": "textStart", "index": idx]

    case .textDelta(let idx, let delta, _):
        dict = ["type": "textDelta", "index": idx, "delta": delta]

    case .textEnd(let idx, let content, _):
        dict = ["type": "textEnd", "index": idx, "content": content]

    case .thinkingStart(let idx, _):
        dict = ["type": "thinkingStart", "index": idx]

    case .thinkingDelta(let idx, let delta, _):
        dict = ["type": "thinkingDelta", "index": idx, "delta": delta]

    case .thinkingEnd(let idx, let content, _):
        dict = ["type": "thinkingEnd", "index": idx, "content": content]

    case .toolCallStart(let idx, _):
        dict = ["type": "toolCallStart", "index": idx]

    case .toolCallDelta(let idx, let delta, _):
        dict = ["type": "toolCallDelta", "index": idx, "delta": delta]

    case .toolCallEnd(let idx, let tc, _):
        dict = [
            "type": "toolCallEnd",
            "index": idx,
            "id": tc.id,
            "name": tc.name,
            "arguments": tc.arguments,
        ]

    case .done(let msg):
        dict = encodeMessage(type: "done", msg: msg)

    case .error(let msg):
        dict = encodeMessage(type: "error", msg: msg)
    }

    return jsonLine(dict)
}

private func encodeMessage(type: String, msg: AssistantMessage) -> [String: Any] {
    var d: [String: Any] = [
        "type":       type,
        "model":      msg.model,
        "provider":   msg.provider,
        "stopReason": msg.stopReason.rawValue,
        "usage": [
            "input":  msg.usage.input,
            "output": msg.usage.output,
            "cost":   msg.usage.cost.total,
        ],
    ]
    if let err = msg.errorMessage { d["errorMessage"] = err }
    if !msg.diagnostics.isEmpty {
        d["diagnostics"] = msg.diagnostics.map { diag -> [String: Any] in
            var entry: [String: Any] = ["type": diag.type]
            if let e = diag.error {
                var errDict: [String: Any] = ["message": e.message]
                if let name = e.name { errDict["name"] = name }
                if let code = e.code { errDict["code"] = code }
                entry["error"] = errDict
            }
            if !diag.details.isEmpty { entry["details"] = diag.details }
            return entry
        }
    }
    return d
}

private func jsonLine(_ dict: [String: Any]) -> String? {
    guard let data = try? JSONSerialization.data(
        withJSONObject: dict,
        options: [.sortedKeys]
    ) else { return nil }
    return String(data: data, encoding: .utf8)
}

// MARK: - Helpers

private func flushStdout() {
    fflush(stdout)
}

func stdinIsTTY() -> Bool {
    isatty(FileHandle.standardInput.fileDescriptor) != 0
}

func readStdin() -> String {
    var lines: [String] = []
    while let line = readLine(strippingNewline: false) {
        lines.append(line)
    }
    return lines.joined().trimmingCharacters(in: .newlines)
}

File: wisp/wisp/output.go
package main

import (
	"encoding/json"
	"fmt"
	"os"

	"wispai"
)

// RunTextMode streams text deltas to stdout.
// Returns (exitCode, finalMessage).
func RunTextMode(ch <-chan wispai.AssistantMessageEvent) (int, *wispai.AssistantMessage) {
	hadText := false
	for event := range ch {
		switch event.Type {
		case wispai.EventTextDelta:
			fmt.Print(event.Delta)
			hadText = true
		case wispai.EventThinkingDelta:
			// Print thinking to stderr so it doesn't pollute stdout pipelines.
			fmt.Fprint(os.Stderr, event.Delta)
		case wispai.EventDone:
			if hadText {
				fmt.Println()
			}
			return 0, event.Message
		case wispai.EventError:
			msg := event.Message
			desc := "unknown error"
			if msg.ErrorMessage != "" {
				desc = msg.ErrorMessage
			}
			fmt.Fprintf(os.Stderr, "\nError: %s\n", desc)
			if len(msg.Diagnostics) > 0 {
				d := msg.Diagnostics[0]
				diagMsg := ""
				if d.Error != nil {
					diagMsg = d.Error.Message
				}
				fmt.Fprintf(os.Stderr, "  [%s] %s\n", d.Type, diagMsg)
			}
			return 1, msg
		}
	}
	return 0, nil
}

// RunJSONMode emits one JSON object per line for every event.
// Returns (exitCode, finalMessage).
func RunJSONMode(ch <-chan wispai.AssistantMessageEvent) (int, *wispai.AssistantMessage) {
	exitCode := 0
	var finalMsg *wispai.AssistantMessage
	for event := range ch {
		if line := encodeEvent(event); line != "" {
			fmt.Println(line)
		}
		switch event.Type {
		case wispai.EventDone:
			finalMsg = event.Message
		case wispai.EventError:
			exitCode = 1
			finalMsg = event.Message
		}
	}
	return exitCode, finalMsg
}

func encodeEvent(e wispai.AssistantMessageEvent) string {
	var d map[string]any
	switch e.Type {
	case wispai.EventStart:
		d = map[string]any{"type": "start", "model": e.Partial.Model, "provider": e.Partial.Provider}
	case wispai.EventTextStart:
		d = map[string]any{"type": "textStart", "index": e.Index}
	case wispai.EventTextDelta:
		d = map[string]any{"type": "textDelta", "index": e.Index, "delta": e.Delta}
	case wispai.EventTextEnd:
		d = map[string]any{"type": "textEnd", "index": e.Index, "content": e.Content}
	case wispai.EventThinkingStart:
		d = map[string]any{"type": "thinkingStart", "index": e.Index}
	case wispai.EventThinkingDelta:
		d = map[string]any{"type": "thinkingDelta", "index": e.Index, "delta": e.Delta}
	case wispai.EventThinkingEnd:
		d = map[string]any{"type": "thinkingEnd", "index": e.Index, "content": e.Content}
	case wispai.EventToolCallStart:
		d = map[string]any{"type": "toolCallStart", "index": e.Index}
	case wispai.EventToolCallDelta:
		d = map[string]any{"type": "toolCallDelta", "index": e.Index, "delta": e.Delta}
	case wispai.EventToolCallEnd:
		d = map[string]any{
			"type": "toolCallEnd", "index": e.Index,
			"id": e.Part.ID, "name": e.Part.Name, "arguments": e.Part.Arguments,
		}
	case wispai.EventDone:
		d = encodeMessage("done", e.Message)
	case wispai.EventError:
		d = encodeMessage("error", e.Message)
	default:
		return ""
	}
	b, _ := json.Marshal(d)
	return string(b)
}

func encodeMessage(typ string, msg *wispai.AssistantMessage) map[string]any {
	d := map[string]any{
		"type":       typ,
		"model":      msg.Model,
		"provider":   msg.Provider,
		"stopReason": string(msg.StopReason),
		"usage": map[string]any{
			"input":  msg.Usage.Input,
			"output": msg.Usage.Output,
			"cost":   msg.Usage.Cost.Total,
		},
	}
	if msg.ErrorMessage != "" {
		d["errorMessage"] = msg.ErrorMessage
	}
	if len(msg.Diagnostics) > 0 {
		diags := make([]map[string]any, 0, len(msg.Diagnostics))
		for _, diag := range msg.Diagnostics {
			de := map[string]any{"type": diag.Type}
			if diag.Error != nil {
				e := map[string]any{"message": diag.Error.Message}
				if diag.Error.Name != "" {
					e["name"] = diag.Error.Name
				}
				if diag.Error.Code != "" {
					e["code"] = diag.Error.Code
				}
				de["error"] = e
			}
			if len(diag.Details) > 0 {
				de["details"] = diag.Details
			}
			diags = append(diags, de)
		}
		d["diagnostics"] = diags
	}
	return d
}
